horovod/common/ops/nccl_operations.cc

```diff
@@ -0,0 +1,294 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include "nccl_operations.h"
+
+namespace horovod {
+namespace common {
+
+ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
+  switch (tensor->dtype()) {
+    case HOROVOD_INT32:
+      return ncclInt32;
+    case HOROVOD_INT64:
+      return ncclInt64;
+    case HOROVOD_FLOAT16:
+      return ncclFloat16;
+    case HOROVOD_FLOAT32:
+      return ncclFloat32;
+    case HOROVOD_FLOAT64:
+      return ncclFloat64;
+    default:
+      throw std::logic_error("Type " + DataType_Name(tensor->dtype()) +
+                             " is not supported in NCCL mode.");
+  }
+}
+
+void NCCLContext::ErrorCheck(std::string op_name, ncclResult_t nccl_result) {
+  if (nccl_result != ncclSuccess) {
+    throw std::logic_error(std::string(op_name) + " failed: " + ncclGetErrorString(nccl_result));
+  }
+}
+
+NCCLAllreduce::NCCLAllreduce(NCCLContext* nccl_context,
+                             MPIContext* mpi_context,
+                             CUDAContext* cuda_context,
+                             HorovodGlobalState* global_state)
+    : CUDAAllreduceAsync(cuda_context, global_state),
+      nccl_context_(nccl_context), mpi_context_(mpi_context) {}
+
+void NCCLAllreduce::InitComm(std::vector<TensorTableEntry>& entries, const std::vector<int32_t>& devices) {
+  // Determine GPU IDs of the devices participating in this communicator.
+  std::vector<int32_t> nccl_device_map = GetDeviceMap(devices);
+
+  // Ensure NCCL communicator is in the map before executing reduction.
+  ncclComm_t& nccl_comm = nccl_context_->nccl_comms[nccl_device_map];
+  if (nccl_comm == nullptr) {
+    auto& timeline = global_state_->timeline;
+    timeline.ActivityStartAll(entries, INIT_NCCL);
+
+    int nccl_rank, nccl_size;
+    Communicator nccl_id_bcast_comm;
+    PopulateCommStrategy(nccl_rank, nccl_size, nccl_id_bcast_comm);
+
+    ncclUniqueId nccl_id;
+    if (nccl_rank == 0) {
+      nccl_context_->ErrorCheck("ncclGetUniqueId", ncclGetUniqueId(&nccl_id));
+    }
+
+    MPI_Broadcast(*mpi_context_, (void*) &nccl_id, sizeof(nccl_id), HOROVOD_BYTE, 0, nccl_id_bcast_comm);
+
+    ncclComm_t new_nccl_comm;
+    auto nccl_result = ncclCommInitRank(&new_nccl_comm, nccl_size, nccl_id, nccl_rank);
```

Comment:
In mpi_operations, you created wrappers for MPI calls, should we align these two approaches?  (Either remove wrappers in MPI, or add here).

Reply:
Removed wrappers.